### DISCUSSION TITLE
Add simulator actions for WriteMultipleRegistersRequest and WriteMultipleCoilsRequest

### DISCUSSION THREAD

**Discussion by krogozinski** [pymodbus-dev/pymodbus#2157] [Ideas]
I have a requirement to define custom actions for `WriteMultipleRegistersRequest` and `WriteMultipleCoilsRequest`. This is for testing the Modbus client on a new hardware project.

To the best of my understanding, simulator actions (custom or built-in) for writes to the simulator server seem to only be performed for `WriteSingleCoilRequest` and `WriteSingleRegisterRequest`, and not `WriteMultipleRegistersRequest` and `WriteMultipleCoilsRequest`.

I can see there is no call to `context.getValues()` in the `WriteMultipleRegistersRequest`  and `WriteMultipleCoilsRequest` `execute()` methods, because it is not strictly required to process the request. Adding this call would allow the actions to be executed.

Is there any appetite for adding this functionality? Am I looking in the right area? I am working in the dev branch.

**Comment by janiversen** [comment]
WriteMultipleCoilsRequest / WriteMultipleRegistersRequest do call context.setValues (in line 230 and 216), why would it call context.getValues() the response do not need the before values (like in the mask call).

A more correct approach would be to expand actions to be called in setValues() as well, that of course would require a parameter change to the action call, e.g. adding the function code.

pull requests are always welcome, if you want something to work differently.

**Reply by krogozinski** [reply]
Understood, thanks for the feedback. I will work on this and submit a PR.

Are you happy for me to convert this discussion to an issue?

**Reply by janiversen** [reply]
ok
